src/index.ts

```diff
@@ -1,87 +1,90 @@
-import { useState, useCallback, useEffect } from "react";
-import update from "immutability-helper";
+import {
+  useMemo,
+  useEffect,
+  useReducer,
+  useCallback,
+} from "react";
 
-import { HookArguments, HookReturnValues, Item } from "./types";
+import reducer from "./reducer";
+import {
+  HookArguments,
+  DefaultItem,
+  ActionType,
+  State,
+} from "./types";
+import { INITIAL_STATE } from "./constants";
 
-export const DEFAULT_ITEM_IDENTIFIER = "id";
-
-function useSelectedItems<T extends Record<any, any>>({
-  items = [],
-  itemIdentifier = DEFAULT_ITEM_IDENTIFIER,
-}: HookArguments<T>): HookReturnValues<T> {
-  const [selectedItems, setSelectedItems] = useState<T[]>([]);
-  const [itemsList, setItemsList] = useState<Item<T>[]>([]);
+function useSelectedItems<T extends DefaultItem, K>({
+  initialItems = [],
+  itemIdentifierKey,
+  initialSelectedItems = [],
+}: HookArguments<T, K>) {
+  const [items, dispatch] = useReducer(
+    reducer,
+    INITIAL_STATE,
+    (state: State) => ({
+      ...state,
+      itemIdentifierKey,
+    }),
+  );
 
   useEffect(() => {
-    const itemIdentifierIsValid = items.some((findItem: T) => (
-      findItem[itemIdentifier]
-    ));
+    const itemsWithInvalidIdentifers = initialItems.filter(
+      (findItem: T) => {
+        const hasItemIdentifierKey = Boolean(findItem[itemIdentifierKey]);
 
-    if (items.length > 0 && !itemIdentifierIsValid) {
-      throw new Error("Please, make sure to provide a valid identifier");
-    }
-  }, [
-    itemIdentifier,
-    items,
-  ]);
+        return !hasItemIdentifierKey;
+      },
+    );
 
-  useEffect(() => {
-    let updatedItemsList: Item<T>[] = [];
-
-    if (items) {
-      updatedItemsList = items.map(listItem => {
-        const isSelected = !!(selectedItems.find(selectedItem => (
-          selectedItem[itemIdentifier] === listItem[itemIdentifier]
-        )));
+    const hasInitialItems = initialItems.length > 0;
+    const hasItemsWithInvalidIdentifersKeys = itemsWithInvalidIdentifers.length > 0;
 
-        return {
-          ...listItem,
-          selected: isSelected,
-        };
-      });
+    if (hasInitialItems && hasItemsWithInvalidIdentifersKeys) {
+      throw new Error(
+        "Please, make sure to provide a valid identifier key to all items",
+      );
     }
-
-    setItemsList(updatedItemsList);
   }, [
-    selectedItems,
-    itemIdentifier,
-    items,
+    itemIdentifierKey,
+    initialItems,
   ]);
 
-  const toggleItem = useCallback((item) => {
-    const selectedItemIndex = selectedItems.findIndex(findItem => (
```

Comment:
I not sure if the new changes with reducer will probably handle some problems here, but before if you call toggleItem twice in the same render phase, you have problems with stale closure, like if I want a toggleAll and implements using toggleItem by all the items, it doesn't work properly because of this stale closure (selectedItems).

The oldest version needed to be something like that to works probably with multiple `toggleItem`:

```tsx
setSelectedItems(prev => {
  const selectedItemIndex = prev.findIndex(findItem => (
    findItem[itemIdentifier] === item[itemIdentifier]
  ));

  const isSelected = selectedItemIndex >= 0;

  if (isSelected) {
    return (
      update(prev, {
        $splice: [[selectedItemIndex, 1]],
      })
    );
  }
  return (
    update(prev, {
      $push: [item],
    })
  );
});
```

I think that with reducer will work, but would be nice to have this feature (toggleAll) in the example to cover this case or some automated tests.

Reply:
The reason why I've changed to the reducer approach is because the hook was too complex, you can found more about it in the issue description. It's not a good practice to have multiple setters calls in different handlers because you have to look to every single block of code to understand the state transitions.

Reply:
There are two ways here: 
 - toggleItems([item1]): Wouldn't be necessary to create another reducer type to handle multiple items 
 - toggleAll: Toggle all the current selected items

Reply:
Can you give more examples related to the stale closure issue?

Reply:
I think it's OK to have multiple setters call, like if you don't expose `toggleAll/toggleItems`, who consumes the library can't use these features because of the stale closure, but if you deal if this, who use your library can make something just like:

```tsx
const toggleItems = (items) => items.forEach(toggleItem)
const toggleAll = () => toggleItems(allItems)
```

Would be nice to have built-in to better optimizations, but if your setters don't permit compose other setters, every edge case needs to be cover by the library.

Reply:
> who consumes the library can't use these features 
Is possible to return the action dispatcher itself or inside another handler to receive arguments

Reply:
Oh, yes, export the action dispatcher is a solution for make the library more composable. I talking more about use the same setter multiple times in the same rendering closure/render phase and makes this setter composable with itself. Like, what I expected if I call `toggleItem` twice with the same `item`: changes to true and after changes to false. But if we have closure state, will only changes to true (twice).

An example when this makes sense: https://twitter.com/diegohaz/status/1288087268222214144

Creates a tutorial like:

```tsx
const runTutorial = async () => {
  toggleItem(item)
  // more code
  await time(2000)
  toggleItem(item)
}
```

or something like:

```tsx
const handleToggle = async (item) => () => {
  toggleItem(item)

  const confirmed = await confirmDialog('Are you sure you want to enable this feature? Etc, etc, etc...')

  if (!confirmed) {
    toggleItem(item)
  }
}

Reply:
And I think now with reducers the stale closure isn't a problem. My main point are about that would be nice to have an example of this, just an idea.

Reply:
> Can you give more examples related to the stale closure issue?

Oh, sorry, I didn't saw that before. Example: https://codesandbox.io/s/confident-tharp-t8h9h?file=/pages/index.tsx

Click twice in `Toggle all` to see the problem.

Reply:
Add a new action called `toggleAllItems`` on [`b2e7ad0` (#11)](https://github.com/LauraBeatris/use-selected-items-hook/pull/11/commits/b2e7ad0b9c87c31fabdfbd3d4e80cb79e838e719)